internal/apis/certmanager/types_certificate.go

```diff
@@ -203,6 +203,25 @@ type CertificatePrivateKey struct {
 	// and will default to `256` if not specified.
 	// No other values are allowed.
 	Size int
+
+	// Other formats in which the key should be written to the secret
+	AdditionalKeyFormats []AdditionalKeyFormat
+}
+
+type KeyFormat string
+
+const (
+	// DER Key format is the binary format of the key
+	// stored in `tls.der` secret key
+	DERKeyFormat KeyFormat = "DER"
+
+	// PEM Key format is a file containing both tls.crt and tls.key
+	// stored in `tls.pem` secret key
+	PEMKeyFormat KeyFormat = "PEM"
```

Comment:
suggestion: I think it'd be best if we keep this PR _only_ for adding the DER-encoded private key. The combined PEM is a good idea which I'd support, but I think it's best to minimise the change in any given PR to make it easier to review.

I think adding the combined PEM in a separate PR would be quick and painless when / if this PR merges.

Reply:
As discussed on slack the feature is now flagged.
So it should not impact users to add those two additional Key Formats.
Also the PEM formating is a very small part of this PR :sweat_smile:

Reply:
This has actually raised a few questions for me:

1) The field is called `additionaKeyFormats`, however, the PEM option here encodes a certificate as well, so 'key' here isn't correct/clear.

2) It _is_ possible to encode a certificate in DER format too, not just a private key (but we can't concatenate them into a single file as _that_ is part of the PEM spec, not DER). If users want DER private keys, is it reasonable to expect they want DER certificate data?

3) Should we therefore move this to `spec.additionalOutputFormats`?

4) If so, and we _do_ include a `CombinedPEM` option, would JKS keystores or PKCS#12 keystores also be configured in a similar way?

I think we need to reconsider how we're expressing these options to make sure we create an API that is cohesive/consistent between different formats (and clear to users when they use it)

cc @SgtCoDFish @maelvls @JoshVanL @irbekrm

Reply:
Your questions are super insightful @munnerz !

There's actually a bunch of overlap here with the `spec.keystores` field on the Certificate resource... so much so, that I can't really think of a reason that we couldn't add DER keys to `keystores` (i.e. add a new field [here](https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1.CertificateKeystores)). Combined chain + key PEM files would also fit under there.

I think both "DER private key" and "combined PEM" would fit under the name "keystores" too. A DER-encoded certificate probably wouldn't fit under "keystores", but I think we could kick that can down the road since nobody's asking for that.

For example:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  ...
spec:
  ...
  keystores:
    derPrivateKey:
      create: true
      # can't think of any configurable options we might want here but we can add later if wanted
    combinedPEM: # combinedPEM just an example, I still don't think we should do this in this PR
      create: true
      ordering: enum(chainFirst | keyFirst) # an example of a configurable option we might want
  ...
```

Reply:
hmm, I disagree - combinedPEM is not a type of keystore, and at least when that field was added it was intended specifically because PKCS#12 and JKS are specifically seen as 'keystores' in common language in the area.

I was more inclined to move `keystores` to be a type of option under `spec.additionalOutputFormats`, e.g.:

```
spec:
  additionalOutputFormats:
  - type: JKS
    passwordSecretRef: a-secret
  - type: DER
  - type: PKCS12
    passwordSecretRef: pkcs12-password
  - type: CombinedPEM
```

Reply:
(but I do agree with your other points regarding this being closely related! 😄)

Reply:
> I was more inclined to move keystores to be a type of option under spec.additionalOutputFormats...

Sure, in an ideal world I fully agree; in practice though this also requires either a migration / deprecation process or for us to support JKS/PKCS12 keystores being requested in two separate places, while also requiring us to write more code to support and test an entirely new field with a new syntax.

I don't think any reasonable person is going to struggle with accepting "keystores" being defined as "any format where a private key is stored"

Under that definition, we wouldn't even need to update the doc comments for `CertificateKeystores` to add both DER and "combined-PEM":

> CertificateKeystores configures additional keystore output formats to be created in the Certificate’s output Secret

Reply:
> I am much more in favor of the additionalOutputFormats option which 1. clearly describes what's going on ("I want additional output formats with my signed certificate chain/private key data/certificate authority data"), and 2. gives scope for adding additional fields to each element, or weird and wacky output formats users might want to use in the future.

In an ideal world I agree and I'd agree for `Certificate` v2 if it's ever created. I still think keystore would be far less confusing for people than having a different field, so that we don't end up with two different fields on the `Certificate` resource which both mean "add some data to my secret under a different key".

I think anything containing a private key could reasonably be thought of as a keystore. JKS and PKCS12 files both contain certificates and private keys and we don't have a problem calling them keystores. If the request was just for a PEM _cert_ with no key I could see why it would be confusing.

That said this _really_ doesn't matter and endlessly bikeshedding API design is a surefire way to never accomplish anything. If adding a new field to `v1.Certificate` will get this (really basic) feature merged then I'll support adding `additionalOutputFormats` [1]. 

[1] (but I wouldn't support then adding JKS and PKCS12 to `additionalOutputFormats` down the road for `v1.Certificate`, that creates two ways of achieving identical outcomes, doubling testing and code attack surface.)

Reply:
@seuf 

In addition, I think it bears repeating here that making this PR _only_ do DER keys would make it much easier to merge, which was the purpose of my original comment on this thread:

> suggestion: I think it'd be best if we keep this PR only for adding the DER-encoded private key. The combined PEM is a good idea which I'd support, but I think it's best to minimise the change in any given PR to make it easier to review.

Reply:
> That said this really doesn't matter and endlessly bikeshedding API design is a surefire way to never accomplish anything.

Without going too off-topic on this PR, I somewhat disagree with this... the API is the way the user interacts with our software, and it is a difficult part to _change_ once we've put it out to the world. Feature gating does help, but historically I don't think we've got a good track record of ownership on a feature-gated feature and we've not been so good at reflecting & iterating on the design later down the road.

I think it's good for us to have this conversation, and especially _with_ the PR author, so that we all have a shared/common understanding of what we're trying to achieve and the longer-term plan for the feature's graduation :)

But I don't want to drag this thread aside from the point at hand, so let's chat on Slack if we want to follow up/bikeshed on our bikeshedding ;)

Reply:
I'll rephrase it:

I think adding this to `keystores` would be better, you both disagree with me. It really, _really_ doesn't matter in the grand scheme of things [1] so let's add `additionalOutputFormats` rather than have me spend time advocating for `keystores`. I'm saying I'm not going to be a blocker; I want to get something done.

[1] The finer points of API design aren't lost on me and I'm well aware of the difficulties of changing a public API but we're just discussing whether to refer to `.spec.keystores.der` or `spec.additionalOutputFormats[].type = der`, it's two different ways of naming the new thing we're going to add.